src/entity.ts

```diff
@@ -0,0 +1,62 @@
+import { Observable } from "rxjs"
+import { first } from 'rxjs/operators'
+import { IApolloQueryOptions, Address } from "./types"
+import { Arc } from "./arc"
+
+// Vote, Competition, Member, Vote, Stake, Plugin, Proposal, DAO, Tag, Event, Reward
+
+interface IEntityState {
+  id: string
+}
+
+export interface IEntityRef<Entity> {
+  id: Address,
+  entity: Entity
+}
+
+export abstract class Entity<TEntityState extends IEntityState> {
+  id: string
+  context: Arc
+  coreState: TEntityState | undefined
+
+  constructor(
+    context: Arc,
+    idOrOpts: string|TEntityState,
+  ) {
+    this.context = context
+    if (typeof idOrOpts === 'string') {
+      this.id = idOrOpts
+    } else {
+      this.id = idOrOpts.id
+      this.setState(idOrOpts)
```

Comment:
Nice! Let's be sure to remove the constructor definitions in derived classes that have this same code, that way there's no boilerplate and easy to see what classes have custom constructor logic (hopefully none :P)